cranelift/codegen/src/loop_analysis.rs

```diff
@@ -198,21 +198,22 @@ impl LoopAnalysis {
         domtree: &DominatorTree,
         layout: &Layout,
     ) {
-        // We traverse the CFG in reverse postorder
-        for &block in domtree.cfg_postorder().iter().rev() {
-            for BlockPredecessor {
-                inst: pred_inst, ..
-            } in cfg.pred_iter(block)
-            {
-                // If the block dominates one of its predecessors it is a back edge
-                if domtree.dominates(block, pred_inst, layout) {
-                    // This block is a loop header, so we create its associated loop
-                    let lp = self.loops.push(LoopData::new(block, None));
-                    self.block_loop_map[block] = lp.into();
-                    break;
-                    // We break because we only need one back edge to identify a loop header.
-                }
-            }
+        for &block in domtree
+            .cfg_postorder()
+            .into_iter()
+            .rev() // We traverse the CFG in reverse postorder
+            .filter(|&&block| {
+                // If the block dominates any one of its predecessors it is a back edge
+                cfg.pred_iter(block).any(
+                    |BlockPredecessor {
+                         inst: pred_inst, ..
+                     }| domtree.dominates(block, pred_inst, layout),
+                )
+            })
```

Comment:
This iterator chain is a little too convoluted to be easily readable, IMHO. One thing that might help is to pull out an `is_loop_header` helper?

It might also be nice to have a method on the domtree that returns an RPO iterator directly (`impl Iterator...` in return is fine). Then we could do something like:

`for &block in domtree.cfg_rpo().filter(|&block| is_loop_header(cfg, domtree, block)) { ... }`